File: app/src/main/java/com/vsa/filmoteca/DetailActivity.java
package com.vsa.filmoteca;

import android.app.AlertDialog;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.app.ActionBarActivity;
import android.view.Menu;
import android.view.MenuItem;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.TextView;

import com.vsa.filmoteca.dialog.DialogManager;
import com.vsa.filmoteca.dialog.interfaces.SimpleDialogListener;
import com.vsa.filmoteca.utils.StringUtils;
import com.vsa.filmoteca.webview.ObservableWebView;

import org.apache.http.protocol.HTTP;

import butterknife.ButterKnife;
import butterknife.InjectView;

public class DetailActivity extends ActionBarActivity implements DetailView{

    public static final String EXTRA_DATE="extra_date";
    public static final String EXTRA_TITLE="extra_title";
    public static final String EXTRA_URL="extra_url";

    @InjectView(R.id.swipe_refresh_layout) SwipeRefreshLayout mSwipeRefreshLayout;
    @InjectView(R.id.webview) ObservableWebView mWebView;
    @InjectView(R.id.detalleTitle) TextView mTitle;

    private ProgressDialog mProgressDialog;


    private DetailPresenter mPresenter;

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
    	setContentView(R.layout.detail);
        ButterKnife.inject(this);
        getSupportActionBar().setDisplayShowTitleEnabled(false);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setDefaultDisplayHomeAsUpEnabled(true);


        mPresenter = new DetailPresenterImpl(this);

        initViews();

        mPresenter.loadContent(getIntent().getStringExtra(EXTRA_URL));
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        mPresenter.onNewIntent(intent);
    }

    @Override
    public void initViews() {
        mSwipeRefreshLayout.setOnRefreshListener(mPresenter);
        mSwipeRefreshLayout.setColorSchemeResources(R.color.color_primary_dark,
                R.color.color_accent,
                R.color.color_primary);
        mWebView.setOnScrollChangedCallback(new ObservableWebView.OnScrollChangedCallback() {
            @Override
            public void onScroll(int l, int t) {
                mSwipeRefreshLayout.setEnabled(t == 0);
            }
        });
        mProgressDialog = ProgressDialog.show(this, "",
                getString(R.string.loading), true,false);
        showMovieTitle(getIntent().getStringExtra(EXTRA_TITLE)
                .substring(1));
    }

    @Override
    public void showProgressDialog() {
        mProgressDialog.show();
    }

    @Override
    public void hideProgressDialog() {
        mProgressDialog.dismiss();
    }

    @Override
    public void showMovieTitle(String title) {
        mTitle.setText(title);
    }

    @Override
    public void stopRefreshing() {
        mSwipeRefreshLayout.setRefreshing(false);
    }

    @Override
	public boolean onCreateOptionsMenu(Menu menu){
		super.onCreateOptionsMenu(menu);
		return mPresenter.onCreateOptionsMenu(getMenuInflater(), menu);
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item){
        return mPresenter.onOptionsItemSelected(item);
	}

	public void showInFilmAffinity(){
		String url="";
		String titulo=this.getIntent().getExtras().getString(EXTRA_TITLE);
		titulo=titulo.replace(" ", "+");
		
		//Le quito los acentos
		titulo=StringUtils.removeAccents(titulo);
		url="http://m.filmaffinity.com/es/search.php?stext="+titulo;
		Intent i_filmaffinity = new Intent(Intent.ACTION_VIEW,Uri.parse(url));
		startActivity(i_filmaffinity);
	}

    public void showInBrowser(){
		String url=this.getIntent().getStringExtra(EXTRA_URL);
		Intent i_navegar = new Intent(Intent.ACTION_VIEW,Uri.parse(url));
		startActivity(i_navegar);
	}

	public void showShareDialog(){
		String tituloCmpBtn = getIntent().getStringExtra(EXTRA_TITLE);
		String fechaCmpBtn = getString(R.string.share_date) + ": " + getIntent().getStringExtra(EXTRA_DATE).substring(1);
		String infoCmpBtn = getString(R.string.share_message) + " " +tituloCmpBtn+"\n"+fechaCmpBtn;
		Intent intent=new Intent(Intent.ACTION_SEND);
		intent.setType(HTTP.PLAIN_TEXT_TYPE);
		intent.putExtra(Intent.EXTRA_SUBJECT, getString(R.string.share_subject));
		intent.putExtra(Intent.EXTRA_TEXT, infoCmpBtn);
		
		startActivity(Intent.createChooser(intent, getString(R.string.share)));
	}

    @Override
    public void showTimeOutDialog(){
        DialogManager.showSimpleDialog(this, R.string.timeout_dialog_message, new SimpleDialogListener(){
            @Override
            public void onAccept() {
                finish();
            }
        });
    }

    @Override
    public void setWebViewContent(String html) {
        mWebView.loadDataWithBaseURL(null, html, "text/html", "utf-8",
                "about:blank");
    }

    @Override
    public Context getContext() {
        return this;
    }

    @Override
    public void showAboutUs() {
        Intent acercade=new Intent(this,AboutActivity.class);
        startActivity(acercade);
    }
}

File: app/src/main/java/com/vsa/filmoteca/presentation/view/activity/DetailActivity.kt
package com.vsa.filmoteca.presentation.view.activity

import android.appwidget.AppWidgetManager
import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.Toast
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.vsa.filmoteca.R
import com.vsa.filmoteca.about.presentation.view.AboutDialog
import com.vsa.filmoteca.databinding.ActivityDetailBinding
import com.vsa.filmoteca.presentation.presenter.detail.DetailPresenter
import com.vsa.filmoteca.presentation.view.DetailView
import com.vsa.filmoteca.presentation.view.dialog.DialogManager
import com.vsa.filmoteca.presentation.view.widget.EventsWidgetProvider
import dagger.hilt.android.AndroidEntryPoint
import javax.inject.Inject

@AndroidEntryPoint
class DetailActivity : BaseActivity(), DetailView, SwipeRefreshLayout.OnRefreshListener {

    companion object {
        const val EXTRA_DATE = "extra_date"
        const val EXTRA_TITLE = "extra_title"
        const val EXTRA_URL = "extra_url"

        fun open(context: Context, url: String, title: String, date: String) {
            context.startActivity(Intent(context, DetailActivity::class.java).apply {
                putExtra(EXTRA_URL, url)
                putExtra(EXTRA_TITLE, title)
                putExtra(EXTRA_DATE, date)
            })
        }
    }

    @Inject
    lateinit var presenter: DetailPresenter

    private lateinit var binding: ActivityDetailBinding

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityDetailBinding.inflate(layoutInflater)
        setContentView(binding.root)
        supportActionBar?.setDisplayShowTitleEnabled(false)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        initViews()
        presenter.onCreate(
            intent.getStringExtra(EXTRA_URL) ?: "",
            intent.getStringExtra(EXTRA_TITLE) ?: ""
        )

    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        presenter.onCreate(
            intent.getStringExtra(EXTRA_URL) ?: "",
            intent.getStringExtra(EXTRA_TITLE) ?: ""
        )
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_detail, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.menu_item_share -> {
                presenter.onShareButtonClick()
                return true
            }
            R.id.menu_item_browser -> {
                presenter.onShowInBrowserButtonClick()
                return true
            }
            R.id.menu_item_filmaffinity -> {
                presenter.onFilmAffinitySearchButtonClick()
                return true
            }
            R.id.menu_item_refresh -> {
                presenter.onRefresh()
                return true
            }
            R.id.menu_item_about_us -> {
                presenter.onAboutUsButtonClick()
                return true
            }
            android.R.id.home -> {
                onBackPressed()
                return true
            }
            else -> return false
        }
    }

    override fun onRefresh() {
        presenter.onRefresh()
    }

    private fun initViews() {
        binding.swipeRefreshLayout.setOnRefreshListener(this)
        binding.swipeRefreshLayout.setColorSchemeResources(
            R.color.color_primary_dark,
            R.color.color_accent,
            R.color.color_primary
        )
        binding.webviewMoviePage.onScrollChangedCallback = { _, top, _, _ ->
            binding.swipeRefreshLayout.isEnabled = top == 0
        }
    }

    override fun showContent() {
        binding.wrapperContent.visibility = View.VISIBLE
    }

    override fun hideContent() {
        binding.wrapperContent.visibility = View.GONE
    }

    override fun showMovieTitle(title: String) {
        binding.textViewTitle.text = title
    }

    override fun updateWidget() {
        val intent = Intent(this, EventsWidgetProvider::class.java)
        intent.action = AppWidgetManager.ACTION_APPWIDGET_UPDATE
        // Use an array and EXTRA_APPWIDGET_IDS instead of AppWidgetManager.EXTRA_APPWIDGET_ID,
        // since it seems the onUpdate() is only fired on that:
        val ids = intArrayOf(R.xml.appwidget_info)
        intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, ids)
        sendBroadcast(intent)
    }

    override fun stopRefreshing() {
        binding.swipeRefreshLayout.isRefreshing = false
    }

    override fun launchBrowser(url: String) {
        if (url.isNotEmpty()) {
            val intent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
            try {
                startActivity(intent)
            } catch (e: ActivityNotFoundException) {
                Toast.makeText(this, R.string.no_browser_found, Toast.LENGTH_SHORT).show()
            }
        }
    }

    override fun showErrorNoInternet() {
        DialogManager.showSimpleDialog(
            this,
            R.string.error_no_internet
        ) { dialog -> dialog.dismiss() }
    }

    override fun showShareDialog() {
        val titleShareButton = intent.getStringExtra(EXTRA_TITLE)
        val dateShareButton =
            getString(R.string.share_date) + ": " + intent.getStringExtra(EXTRA_DATE)?.substring(1)
        val infoShareButton =
            getString(R.string.share_message) + " " + titleShareButton + "\n" + dateShareButton
        val intent = Intent(Intent.ACTION_SEND)
        intent.type = "text/plain"
        intent.putExtra(Intent.EXTRA_SUBJECT, getString(R.string.share_subject))
        intent.putExtra(Intent.EXTRA_TEXT, infoShareButton)

        startActivity(Intent.createChooser(intent, getString(R.string.share)))
    }

    override fun showTimeOutDialog() {
        DialogManager.showSimpleDialog(this, R.string.timeout_dialog_message) { finish() }
    }

    override fun setWebViewContent(html: String, baseUrl: String) {
        binding.webviewMoviePage.loadDataWithBaseURL(
            baseUrl, html, "text/html", "utf-8",
            "about:blank"
        )
    }

    override fun showAboutUs() {
        AboutDialog.show(supportFragmentManager)
    }

}
